File: src/core/mp2.ts
/** MP2 (second-order Moller-Plesset perturbation theory) correlation energy.
 *
 *  E_MP2 = Σ_{i<j,a<b} (ia|jb)[2(ia|jb) - (ib|ja)] / (ε_i + ε_j - ε_a - ε_b)
 *
 *  MO ERI via two half-transformations: O(N^5) total.
 *
 *  Step 1 (first half): for each i, build tmp_i[a,λ,σ] = Σ_μν C[μ,i]*C[ν,a]*(μν|λσ)
 *    Decomposed as:
 *      B[μ,a,λ,σ] = Σ_ν C[ν,a+nocc] * (μν|λσ)     ... O(N * nvir * N^2) per μ
 *      tmp_i[a,λ,σ] = Σ_μ C[μ,i] * B[μ,a,λ,σ]      ... O(N * nvir * N^2)
 *    Total step 1: O(nocc * N^2 * N * nvir) = O(N^5)
 *
 *  Step 2 (second half): (ia|jb) = Σ_λσ C[λ,j] * C[σ,b+nocc] * tmp_i[a,λ,σ]
 *    For each (i,j) pair, contract tmp_i and tmp_j to get (ia|jb) for all a,b
 *    Total step 2: O(nocc^2 * nvir^2 * N^2) ≈ O(N^4) to O(N^5)
 */

import type { ERIStored } from './eri';
import { Matrix, type FloatArray } from '../linalg/matrix';
import { matmul, matmulAtB } from '../linalg/matmul';
import { jacobiEigen } from '../linalg/eigendecomposition';

/** Compute MP2 correlation energy from converged RHF results.
 *  @param C       MO coefficient matrix (AO x MO), columns sorted by orbital energy
 *  @param epsilon Orbital energies (sorted)
 *  @param eri     AO basis ERI storage
 *  @param nocc    Number of occupied orbitals
 *  @param nbasis  Number of basis functions
 */
export function computeMP2Energy(
  C: Matrix,
  epsilon: FloatArray,
  eri: ERIStored,
  nocc: number,
  nbasis: number,
): number {
  const nvir = nbasis - nocc;
  const N = nbasis;

  // Step 1: Build half-transformed integrals for each occupied orbital i.
  // half[i] is a Float64Array of size [nvir × N × N], storing tmp_i[a, λ, σ].
  //
  // tmp_i[a,λ,σ] = Σ_μ C[μ,i] * Σ_ν C[ν,a+nocc] * (μν|λσ)
  //
  // We decompose into two steps to avoid O(N^4) per (i,a,λ,σ):
  //   For each (λ,σ): buf[μ,a] = Σ_ν C[ν,a+nocc] * (μν|λσ)  ... O(N^2 * nvir * N) total over all λ,σ
  //                    tmp_i[a,λ,σ] = Σ_μ C[μ,i] * buf[μ,a]   ... O(nocc * nvir * N) per (λ,σ)

  const halfTransformed: Float64Array[] = new Array(nocc);
  const buf = new Float64Array(N * nvir); // buf[mu * nvir + a]

  for (let i = 0; i < nocc; i++) {
    const tmp = new Float64Array(nvir * N * N); // tmp[a * N*N + lam * N + sig]
    halfTransformed[i] = tmp;

    for (let lam = 0; lam < N; lam++) {
      for (let sig = 0; sig < N; sig++) {
        // First quarter-transform: buf[mu, a] = Σ_ν C[ν, a+nocc] * (μν|λσ)
        buf.fill(0);
        for (let mu = 0; mu < N; mu++) {
          for (let nu = 0; nu < N; nu++) {
            const eriVal = eri.get(mu, nu, lam, sig);
            if (Math.abs(eriVal) < 1e-15) continue;
            for (let a = 0; a < nvir; a++) {
              buf[mu * nvir + a] += C.get(nu, nocc + a) * eriVal;
            }
          }
        }

        // Second quarter-transform: tmp[a, lam, sig] = Σ_μ C[μ,i] * buf[μ, a]
        for (let mu = 0; mu < N; mu++) {
          const Cmi = C.get(mu, i);
          if (Math.abs(Cmi) < 1e-15) continue;
          for (let a = 0; a < nvir; a++) {
            tmp[a * N * N + lam * N + sig] += Cmi * buf[mu * nvir + a];
          }
        }
      }
    }
  }

  // Step 2: Contract half-transformed integrals to get (ia|jb) and accumulate MP2 energy.
  // (ia|jb) = Σ_λσ C[λ,j] * C[σ,b+nocc] * half[i][a, λ, σ]
  //
  // Optimization: precompute second half for j as well:
  //   half[j][b, λ, σ] is already computed, so:
  //   (ia|jb) = Σ_λσ half_i[a,λ,σ] * ... but half_i already has the bra indices transformed.
  //   We need: (ia|jb) = Σ_λσ tmp_i[a,λ,σ] * C[λ,j] * C[σ,b+nocc]
  //
  // Actually, tmp_i[a,λ,σ] has bra fully transformed (i,a) and ket in AO (λ,σ).
  // So (ia|jb) = Σ_λ C[λ,j] * Σ_σ C[σ,b+nocc] * tmp_i[a,λ,σ]

  let emp2 = 0;

  for (let i = 0; i < nocc; i++) {
    const tmp_i = halfTransformed[i];
    for (let j = i; j < nocc; j++) {
      // For this (i,j) pair, compute all (ia|jb) and (ib|ja)
      for (let a = 0; a < nvir; a++) {
        for (let b = 0; b < nvir; b++) {
          // (ia|jb) = Σ_λσ C[λ,j] * C[σ,b+nocc] * tmp_i[a,λ,σ]
          let iajb = 0;
          for (let lam = 0; lam < N; lam++) {
            const Clj = C.get(lam, j);
            if (Math.abs(Clj) < 1e-15) continue;
            for (let sig = 0; sig < N; sig++) {
              iajb += Clj * C.get(sig, nocc + b) * tmp_i[a * N * N + lam * N + sig];
            }
          }

          // (ib|ja) = Σ_λσ C[λ,j] * C[σ,a+nocc] * tmp_i[b,λ,σ]
          // Note: tmp_i[b,...] already has the i index baked in
          let ibja = 0;
          for (let lam = 0; lam < N; lam++) {
            const Clj = C.get(lam, j);
            if (Math.abs(Clj) < 1e-15) continue;
            for (let sig = 0; sig < N; sig++) {
              ibja += Clj * C.get(sig, nocc + a) * tmp_i[b * N * N + lam * N + sig];
            }
          }

          const aa = nocc + a;
          const bb = nocc + b;
          const denom = epsilon[i] + epsilon[j] - epsilon[aa] - epsilon[bb];
          const factor = (i === j) ? 1.0 : 2.0; // account for i<j symmetry
          emp2 += factor * iajb * (2 * iajb - ibja) / denom;
        }
      }
    }
  }

  return emp2;
}

// ---------------------------------------------------------------------------
// UMP2 / ROMP2
// ---------------------------------------------------------------------------

/** Half-transform AO ERIs with coefficients Cocc (occupied) and Cvir (virtual).
 *  Returns half[i][a, λ, σ] for each occupied orbital i.
 *  bra is transformed: (i_occ, a_vir | λ_AO, σ_AO)
 */
function halfTransform(
  Cocc: Matrix, noccOcc: number,
  Cvir: Matrix, noccVir: number, nvirVir: number,
  eri: ERIStored, nbasis: number,
): Float64Array[] {
  const N = nbasis;
  const half: Float64Array[] = new Array(noccOcc);
  const buf = new Float64Array(N * nvirVir);

  for (let i = 0; i < noccOcc; i++) {
    const tmp = new Float64Array(nvirVir * N * N);
    half[i] = tmp;

    for (let lam = 0; lam < N; lam++) {
      for (let sig = 0; sig < N; sig++) {
        buf.fill(0);
        for (let mu = 0; mu < N; mu++) {
          for (let nu = 0; nu < N; nu++) {
            const eriVal = eri.get(mu, nu, lam, sig);
            if (Math.abs(eriVal) < 1e-15) continue;
            for (let a = 0; a < nvirVir; a++) {
              buf[mu * nvirVir + a] += Cvir.get(nu, noccVir + a) * eriVal;
            }
          }
        }
        for (let mu = 0; mu < N; mu++) {
          const Cmi = Cocc.get(mu, i);
          if (Math.abs(Cmi) < 1e-15) continue;
          for (let a = 0; a < nvirVir; a++) {
            tmp[a * N * N + lam * N + sig] += Cmi * buf[mu * nvirVir + a];
          }
        }
      }
    }
  }
  return half;
}

/** Contract half-transformed integral (ia|λσ) with Cj and Cb to get (ia|jb).
 *  Returns the integral value.
 */
function contractMOInt(
  halfI: Float64Array, a: number,
  Cj: Matrix, j: number,
  Cb: Matrix, bOffset: number, b: number,
  N: number,
): number {
  let val = 0;
  for (let lam = 0; lam < N; lam++) {
    const clj = Cj.get(lam, j);
    if (Math.abs(clj) < 1e-15) continue;
    for (let sig = 0; sig < N; sig++) {
      val += clj * Cb.get(sig, bOffset + b) * halfI[a * N * N + lam * N + sig];
    }
  }
  return val;
}

/** Compute UMP2 correlation energy from converged UHF results.
 *  E = E_αα + E_ββ + E_αβ
 */
export function computeUMP2Energy(
  Ca: Matrix, Cb: Matrix,
  epsilonA: FloatArray, epsilonB: FloatArray,
  eri: ERIStored,
  noccA: number, noccB: number,
  nbasis: number,
): number {
  const N = nbasis;
  const nvirA = N - noccA;
  const nvirB = N - noccB;

  // Half-transform: bra=(i_α, a_α), ket=AO
  const halfA = halfTransform(Ca, noccA, Ca, noccA, nvirA, eri, N);
  // Half-transform: bra=(i_β, a_β), ket=AO
  const halfB = halfTransform(Cb, noccB, Cb, noccB, nvirB, eri, N);

  let emp2 = 0;

  // αα contribution: E = Σ_{i<j} Σ_{a<b} |(ia|jb)-(ib|ja)|² / denom
  for (let i = 0; i < noccA; i++) {
    for (let j = i + 1; j < noccA; j++) {
      for (let a = 0; a < nvirA; a++) {
        for (let b = a + 1; b < nvirA; b++) {
          const iajb = contractMOInt(halfA[i], a, Ca, j, Ca, noccA, b, N);
          const ibja = contractMOInt(halfA[i], b, Ca, j, Ca, noccA, a, N);
          const antisym = iajb - ibja;
          const denom = epsilonA[i] + epsilonA[j] - epsilonA[noccA + a] - epsilonA[noccA + b];
          emp2 += antisym * antisym / denom;
        }
      }
    }
  }

  // ββ contribution
  for (let i = 0; i < noccB; i++) {
    for (let j = i + 1; j < noccB; j++) {
      for (let a = 0; a < nvirB; a++) {
        for (let b = a + 1; b < nvirB; b++) {
          const iajb = contractMOInt(halfB[i], a, Cb, j, Cb, noccB, b, N);
          const ibja = contractMOInt(halfB[i], b, Cb, j, Cb, noccB, a, N);
          const antisym = iajb - ibja;
          const denom = epsilonB[i] + epsilonB[j] - epsilonB[noccB + a] - epsilonB[noccB + b];
          emp2 += antisym * antisym / denom;
        }
      }
    }
  }

  // αβ contribution: E = Σ_{iα,jβ,aα,bβ} |(iα aα|jβ bβ)|² / denom
  // half_iα has bra=(iα, aα) with Cα, ket still in AO.
  // Need to contract ket with Cβ for j and b.
  for (let i = 0; i < noccA; i++) {
    for (let j = 0; j < noccB; j++) {
      for (let a = 0; a < nvirA; a++) {
        for (let b = 0; b < nvirB; b++) {
          const iajb = contractMOInt(halfA[i], a, Cb, j, Cb, noccB, b, N);
          const denom = epsilonA[i] + epsilonB[j] - epsilonA[noccA + a] - epsilonB[noccB + b];
          emp2 += iajb * iajb / denom;
        }
      }
    }
  }

  return emp2;
}

/** Compute ROMP2 (semicanonical) correlation energy from converged ROHF results.
 *  1. Build semicanonical orbitals by block-diagonalizing Fα and Fβ in MO basis
 *  2. Call UMP2 with the resulting coefficients and orbital energies
 */
export function computeROMP2Energy(
  C: Matrix,
  Fa: Matrix, Fb: Matrix,
  eri: ERIStored,
  noccA: number, noccB: number,
  nbasis: number,
): number {
  const N = nbasis;

  // Transform Fock matrices to MO basis: F_MO = C^T F C
  const FaMO = matmul(matmulAtB(C, Fa), C);
  const FbMO = matmul(matmulAtB(C, Fb), C);

  // 2-block semicanonical: α occ=[0,noccA) vir=[noccA,N), β occ=[0,noccB) vir=[noccB,N)
  function semicanonicalize(FMO: Matrix, nocc: number): { U: Matrix; epsilon: Float64Array } {
    const nvir = N - nocc;
    const U = Matrix.identity(N);
    const eps = new Float64Array(N);

    // Occupied block
    if (nocc > 0) {
      const occ = new Matrix(nocc, nocc);
      for (let i = 0; i < nocc; i++)
        for (let j = 0; j < nocc; j++)
          occ.set(i, j, FMO.get(i, j));
      const eig = jacobiEigen(occ);
      for (let i = 0; i < nocc; i++) {
        eps[i] = eig.eigenvalues[i];
        for (let j = 0; j < nocc; j++)
          U.set(i, j, eig.eigenvectors.get(i, j));
      }
    }

    // Virtual block
    if (nvir > 0) {
      const vir = new Matrix(nvir, nvir);
      for (let i = 0; i < nvir; i++)
        for (let j = 0; j < nvir; j++)
          vir.set(i, j, FMO.get(nocc + i, nocc + j));
      const eig = jacobiEigen(vir);
      for (let i = 0; i < nvir; i++) {
        eps[nocc + i] = eig.eigenvalues[i];
        for (let j = 0; j < nvir; j++)
          U.set(nocc + i, nocc + j, eig.eigenvectors.get(i, j));
      }
    }

    return { U, epsilon: eps };
  }

  const { U: Ua, epsilon: epsA } = semicanonicalize(FaMO, noccA);
  const { U: Ub, epsilon: epsB } = semicanonicalize(FbMO, noccB);

  // Semicanonical coefficients: C_sc = C * U
  const CaScf = matmul(C, Ua);
  const CbScf = matmul(C, Ub);

  return computeUMP2Energy(CaScf, CbScf, epsA, epsB, eri, noccA, noccB, N);
}

File: wasm-eri/src/mp2.rs
/// WASM-accelerated MP2 correlation energy — port of mp2.ts

use crate::integrals::eri_1d_index;
use crate::mp3::full_mo_transform_4c;
use crate::simd_utils;

/// Compute MP2 correlation energy from converged RHF results.
///
/// `eri`: 8-fold symmetric ERI data (flat f64)
/// `coeff`: MO coefficient matrix (N*N, row-major, AO x MO)
/// `epsilon`: orbital energies (N elements, sorted)
/// `nocc`: number of occupied spatial orbitals
/// `n`: number of basis functions
///
/// Algorithm: two half-transforms + contraction, O(N^5) total.
pub fn compute_mp2(
    eri: &[f64],
    coeff: &[f64],
    epsilon: &[f64],
    nocc: usize,
    n: usize,
) -> f64 {
    let nvir = n - nocc;

    // Pre-extract virtual orbital coefficient columns (contiguous for SIMD)
    // c_vir[a][sig] = coeff[sig * n + (nocc + a)]
    let mut c_vir: Vec<Vec<f64>> = Vec::with_capacity(nvir);
    for a in 0..nvir {
        let mut col = vec![0.0_f64; n];
        for sig in 0..n {
            col[sig] = coeff[sig * n + (nocc + a)];
        }
        c_vir.push(col);
    }

    // Pre-extract occupied orbital coefficient columns
    // c_occ[j][lam] = coeff[lam * n + j]
    let mut c_occ: Vec<Vec<f64>> = Vec::with_capacity(nocc);
    for j in 0..nocc {
        let mut col = vec![0.0_f64; n];
        for lam in 0..n {
            col[lam] = coeff[lam * n + j];
        }
        c_occ.push(col);
    }

    // Step 1: Build half-transformed integrals for each occupied orbital i.
    let mut half_transformed: Vec<Vec<f64>> = Vec::with_capacity(nocc);
    let mut buf = vec![0.0_f64; n * nvir];

    for i in 0..nocc {
        let mut tmp = vec![0.0_f64; nvir * n * n];

        for lam in 0..n {
            for sig in 0..n {
                for x in buf.iter_mut() {
                    *x = 0.0;
                }
                for mu in 0..n {
                    for nu in 0..n {
                        let eri_val = eri[eri_1d_index(mu, nu, lam, sig)];
                        if eri_val.abs() < 1e-15 {
                            continue;
                        }
                        // buf[mu, a] += C_vir[a][nu] * eri_val -- use daxpy
                        for a in 0..nvir {
                            buf[mu * nvir + a] += c_vir[a][nu] * eri_val;
                        }
                    }
                }

                for mu in 0..n {
                    let c_mi = c_occ[i][mu];
                    if c_mi.abs() < 1e-15 {
                        continue;
                    }
                    for a in 0..nvir {
                        tmp[a * n * n + lam * n + sig] += c_mi * buf[mu * nvir + a];
                    }
                }
            }
        }

        half_transformed.push(tmp);
    }

    // Step 2: Contract using SIMD dot products.
    // (ia|jb) = Σ_λ C_occ[j][λ] * Σ_σ C_vir[b][σ] * tmp[a, λ, σ]
    //         = Σ_λ C_occ[j][λ] * dot(C_vir[b], tmp[a, λ, :])
    let mut emp2 = 0.0_f64;

    for i in 0..nocc {
        let tmp_i = &half_transformed[i];
        for j in i..nocc {
            for a in 0..nvir {
                for b in 0..nvir {
                    // (ia|jb) = Σ_λ C_occ[j][λ] * dot(C_vir[b], tmp_i[a*n*n + λ*n .. +n])
                    let mut iajb = 0.0_f64;
                    let base_a = a * n * n;
                    for lam in 0..n {
                        let c_lj = c_occ[j][lam];
                        if c_lj.abs() < 1e-15 { continue; }
                        let row_start = base_a + lam * n;
                        iajb += c_lj * simd_utils::dot(&c_vir[b], &tmp_i[row_start..row_start + n]);
                    }

                    // (ib|ja) = Σ_λ C_occ[j][λ] * dot(C_vir[a], tmp_i[b*n*n + λ*n .. +n])
                    let mut ibja = 0.0_f64;
                    let base_b = b * n * n;
                    for lam in 0..n {
                        let c_lj = c_occ[j][lam];
                        if c_lj.abs() < 1e-15 { continue; }
                        let row_start = base_b + lam * n;
                        ibja += c_lj * simd_utils::dot(&c_vir[a], &tmp_i[row_start..row_start + n]);
                    }

                    let aa = nocc + a;
                    let bb = nocc + b;
                    let denom = epsilon[i] + epsilon[j] - epsilon[aa] - epsilon[bb];
                    let factor = if i == j { 1.0 } else { 2.0 };
                    emp2 += factor * iajb * (2.0 * iajb - ibja) / denom;
                }
            }
        }
    }

    emp2
}

/// Compute UMP2 correlation energy from converged UHF results.
///
/// `eri`: 8-fold symmetric ERI data
/// `ca`: alpha MO coefficient matrix (N*N, row-major)
/// `cb`: beta MO coefficient matrix (N*N, row-major)
/// `epsilon_a`: alpha orbital energies (N elements)
/// `epsilon_b`: beta orbital energies (N elements)
/// `nocc_a`: number of occupied alpha orbitals
/// `nocc_b`: number of occupied beta orbitals
/// `n`: number of basis functions
pub fn compute_ump2(
    eri: &[f64],
    ca: &[f64],
    cb: &[f64],
    epsilon_a: &[f64],
    epsilon_b: &[f64],
    nocc_a: usize,
    nocc_b: usize,
    n: usize,
) -> f64 {
    let nvir_a = n - nocc_a;
    let nvir_b = n - nocc_b;
    let n2 = n * n;
    let n3 = n2 * n;

    // Full MO integral transforms for the 3 spin blocks
    let mo_aa = full_mo_transform_4c(ca, ca, ca, ca, eri, n);
    let mo_bb = full_mo_transform_4c(cb, cb, cb, cb, eri, n);
    let mo_ab = full_mo_transform_4c(ca, ca, cb, cb, eri, n);

    let mut emp2 = 0.0_f64;

    // αα contribution: Σ_{i<j}^α Σ_{a<b}^α |(ia|jb)-(ib|ja)|² / D
    for i in 0..nocc_a {
        for j in (i + 1)..nocc_a {
            for a in 0..nvir_a {
                for b in (a + 1)..nvir_a {
                    let aa = nocc_a + a;
                    let bb = nocc_a + b;
                    let iajb = mo_aa[i * n3 + aa * n2 + j * n + bb];
                    let ibja = mo_aa[i * n3 + bb * n2 + j * n + aa];
                    let antisym = iajb - ibja;
                    let denom = epsilon_a[i] + epsilon_a[j] - epsilon_a[aa] - epsilon_a[bb];
                    emp2 += antisym * antisym / denom;
                }
            }
        }
    }

    // ββ contribution: Σ_{i<j}^β Σ_{a<b}^β |(ia|jb)-(ib|ja)|² / D
    for i in 0..nocc_b {
        for j in (i + 1)..nocc_b {
            for a in 0..nvir_b {
                for b in (a + 1)..nvir_b {
                    let aa = nocc_b + a;
                    let bb = nocc_b + b;
                    let iajb = mo_bb[i * n3 + aa * n2 + j * n + bb];
                    let ibja = mo_bb[i * n3 + bb * n2 + j * n + aa];
                    let antisym = iajb - ibja;
                    let denom = epsilon_b[i] + epsilon_b[j] - epsilon_b[aa] - epsilon_b[bb];
                    emp2 += antisym * antisym / denom;
                }
            }
        }
    }

    // αβ contribution: Σ_{iα,jβ,aα,bβ} |(iα aα|jβ bβ)|² / D
    for i in 0..nocc_a {
        for j in 0..nocc_b {
            for a in 0..nvir_a {
                for b in 0..nvir_b {
                    let aa = nocc_a + a;
                    let bb = nocc_b + b;
                    let iajb = mo_ab[i * n3 + aa * n2 + j * n + bb];
                    let denom = epsilon_a[i] + epsilon_b[j] - epsilon_a[aa] - epsilon_b[bb];
                    emp2 += iajb * iajb / denom;
                }
            }
        }
    }

    emp2
}
